File: src/backend/message.ts
/**
 * Copyright 2025 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Serializable } from "@/internals/serializable.js";
import { shallowCopy } from "@/serializer/utils.js";
import { FilePart, ImagePart, TextPart, ToolCallPart, ToolResultPart } from "ai";
import { z } from "zod";
import { ValueError } from "@/errors.js";

export type MessageRole = "user" | "system" | "tool" | "assistant";
export type MessageContentPart = TextPart | ToolCallPart | ImagePart | FilePart | ToolResultPart;

export interface MessageMeta {
  [key: string]: any;
  createdAt?: Date;
}

export interface MessageInput {
  role: MessageRole;
  text: string; // TODO
  meta?: MessageMeta;
}

function isText(content: MessageContentPart): content is TextPart {
  return content.type === "text";
}
function isImage(content: MessageContentPart): content is ImagePart {
  return content.type === "image";
}
function isFile(content: MessageContentPart): content is FilePart {
  return content.type === "file";
}
function isToolCall(content: MessageContentPart): content is ToolCallPart {
  return content.type === "tool-call";
}
function isToolResult(content: MessageContentPart): content is ToolResultPart {
  return content.type === "tool-result";
}

export abstract class Message<
  T extends MessageContentPart = MessageContentPart,
  R extends string = MessageRole | string,
> extends Serializable {
  public abstract readonly role: R;
  public readonly content: T[];

  constructor(
    content: T | T[] | string,
    public readonly meta: MessageMeta = {},
  ) {
    super();
    if (!meta?.createdAt) {
      meta.createdAt = new Date();
    }
    if (typeof content === "string") {
      this.content = [this.fromString(content)];
    } else {
      this.content = Array.isArray(content) ? content : [content];
    }
  }

  protected abstract fromString(input: string): T;

  static of({ role, text, meta }: MessageInput) {
    if (role === "user") {
      return new UserMessage(text, meta);
    } else if (role === "assistant") {
      return new AssistantMessage(text, meta);
    } else if (role === "system") {
      return new SystemMessage(text, meta);
    } else if (role === "tool") {
      return new ToolMessage(text, meta);
    } else {
      return new CustomMessage(role, text, meta);
    }
  }

  get text() {
    return this.getTexts()
      .map((c) => c.text)
      .join("");
  }

  getTexts() {
    return this.content.filter(isText) as TextPart[];
  }

  createSnapshot() {
    return { content: shallowCopy(this.content), meta: shallowCopy(this.meta), role: this.role };
  }

  loadSnapshot(snapshot: ReturnType<typeof this.createSnapshot>) {
    Object.assign(this, snapshot);
  }

  toPlain() {
    return { role: this.role, content: shallowCopy(this.content) } as const;
  }

  [Symbol.iterator](): Iterator<T> {
    return this.content[Symbol.iterator]();
  }
}

export class AssistantMessage extends Message<TextPart | ToolCallPart> {
  public readonly role = "assistant";

  static {
    this.register();
  }

  getToolCalls() {
    return this.content.filter(isToolCall);
  }

  protected fromString(text: string): TextPart {
    return { type: "text", text };
  }
}

export class ToolMessage extends Message<ToolResultPart> {
  public readonly role = "tool";

  static {
    this.register();
  }

  getToolResults() {
    return this.content.filter(isToolResult);
  }

  protected fromString(text: string): ToolResultPart {
    const { success, data } = z
      .object({
        type: z.literal("tool-result"),
        result: z.any(),
        toolName: z.string(),
        toolCallId: z.string(),
      })
      .safeParse(text);

    if (!success) {
      throw new ValueError(`ToolMessage cannot be created from '${text}'!`);
    }

    return data as ToolResultPart;
  }
}

export class SystemMessage extends Message<TextPart> {
  public readonly role: MessageRole = "system";

  static {
    this.register();
  }

  protected fromString(text: string): TextPart {
    return { type: "text", text };
  }
}

export class UserMessage extends Message<TextPart | ImagePart | FilePart> {
  public readonly role = "user";

  static {
    this.register();
  }

  getImages() {
    return this.content.filter(isImage);
  }

  getFiles() {
    return this.content.filter(isFile);
  }

  protected fromString(text: string): TextPart {
    return { type: "text", text };
  }
}

export const Role = {
  ASSISTANT: "assistant",
  SYSTEM: "system",
  USER: "user",
} as const;

export class CustomMessage extends Message<MessageContentPart, string> {
  public role: string;

  constructor(role: string, content: MessageContentPart | string, meta: MessageMeta = {}) {
    super(content, meta);
    if (!role) {
      throw new ValueError(`Role "${role}" must be specified!`);
    }
    this.role = role;
  }

  protected fromString(input: string): MessageContentPart {
    return { type: "text", text: input };
  }
}

File: python/beeai_framework/backend/message.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import json
from abc import abstractmethod
from datetime import UTC, datetime
from enum import Enum
from typing import Any, Generic, Literal, TypeVar

from pydantic import BaseModel, Field

from beeai_framework.backend import MessageError

T = TypeVar("T", bound=BaseModel)
MessageMeta = dict[str, Any]


class Role(str, Enum):
    ASSISTANT: str = "assistant"
    SYSTEM: str = "system"
    TOOL: str = "tool"
    USER: str = "user"

    def __str__(self) -> str:
        return self.value

    @classmethod
    def values(cls) -> set[str]:
        return {value for key, value in vars(cls).items() if not key.startswith("_") and isinstance(value, str)}


class ToolResult(BaseModel):
    type: Literal["tool-result"]
    result: Any
    tool_name: str = Field(alias="toolName")
    tool_call_id: str = Field(alias="toolCallId")


class MessageInput(BaseModel):
    role: Role | str
    text: str
    meta: MessageMeta | None = None


class Message(Generic[T]):
    role: Role | str
    content: T

    def __init__(self, content: T | list[T], meta: MessageMeta | None = None) -> None:
        if meta and not meta.get("createdAt"):
            meta["createdAt"] = datetime.now(tz=UTC)

        if isinstance(content, str):
            self.content = [self.from_string(text=content)]
        elif isinstance(content, list):
            self.content = content
        else:
            self.content = [content]

    @property
    def role(self) -> Role | str:
        return self._role

    @property
    def text(self) -> str:
        return "".join([x.get("text") for x in self.get_texts()])

    @abstractmethod
    def from_string(self, text: str) -> T:
        pass

    def get_texts(self) -> list[T]:
        return list(filter(lambda x: x.get("type") == "text", self.content))

    def to_plain(self) -> dict[str, Any]:
        return {"role": self.role, "content": self.text}

    @classmethod
    def of(cls, message_data: dict[str, str]) -> "Message":
        message_input = MessageInput.model_validate(message_data, strict=True)
        if message_input.role == Role.USER:
            return UserMessage(message_input.text, message_input.meta)
        elif message_input.role == Role.ASSISTANT:
            return AssistantMessage(message_input.text, message_input.meta)
        elif message_input.role == Role.SYSTEM:
            return SystemMessage(message_input.text, message_input.meta)
        elif message_input.role == Role.TOOL:
            return ToolMessage(message_input.text, message_input.meta)
        else:
            return CustomMessage(message_input.role, message_input.text, message_input.meta)


class AssistantMessage(Message):
    _role = Role.ASSISTANT

    def from_string(self, text: str) -> T:
        return {"type": "text", "text": text}

    def get_tool_calls(self) -> list[T]:
        return filter(lambda x: x.get("type") == "tool-call", self.content)


class ToolMessage(Message):
    _role = Role.TOOL

    def from_string(self, text: str) -> ToolResult:
        tool_result = ToolResult.model_validate(json.loads(text))
        return tool_result.model_dump(by_alias=True)

    def get_tool_results(self) -> list[T]:
        return filter(lambda x: x.get("type") == "tool-result", self.content)


class SystemMessage(Message):
    _role = Role.SYSTEM

    def from_string(self, text: str) -> T:
        return {"type": "text", "text": text}


class UserMessage(Message):
    _role = Role.USER

    def from_string(self, text: str) -> T:
        return {"type": "text", "text": text}

    def get_images(self) -> list[T]:
        return filter(lambda x: x.get("type") == "image", self.content)

    def get_files(self) -> list[T]:
        return filter(lambda x: x.get("type") == "file", self.content)


class CustomMessage(Message):
    def __init__(self, role: str, content: T, meta: MessageMeta = None) -> None:
        super().__init__(content, meta)
        if not role:
            raise MessageError("Role must be specified!")
        self._role = role

    def from_string(self, text: str) -> dict[str, Any]:
        return {"type": "text", "text": text}
